src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php

```diff
@@ -184,7 +184,15 @@ private function parseDefinition(\DOMElement $service, $file)
                 } elseif ($childService = $factory->getAttribute('service')) {
                     $class = new Reference($childService, ContainerInterface::EXCEPTION_ON_INVALID_REFERENCE, false);
                 } else {
-                    $class = $factory->getAttribute('class');
+                    $class = $factory->getAttribute('class') ?: $definition->getClass();
+                }
+
+                if (null === $class) {
```

Comment:
Right now, I don't see the benefits, nor how this method would be helpful for this PR. BTW, you still miss information about the place from where was loaded the service.
DX indeed always increase complexity, as it requires writing specific code and sometimes prevent doing wrong things ahead. But I think we're going to survive this extra `if` 😅 

However, if you think you have a point about the `Definition::resolve($id)` suggestion, I think it worth a PR in order to show us the benefits. :)